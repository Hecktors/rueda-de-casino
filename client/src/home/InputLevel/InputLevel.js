import { useState, useEffect } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components/macro'
import { ArrowRightIcon, ArrowDownIcon } from '../../app/components/Icons/Icons'

InputLevel.propTypes = {
  levelName: PropTypes.string.isRequired,
  levelMoves: PropTypes.array.isRequired,
  selectedMoveIDs: PropTypes.array.isRequired,
  updateAppState: PropTypes.func.isRequired,
}

export default function InputLevel({
  levelName,
  levelMoves,
  selectedMoveIDs,
  updateAppState,
}) {
  const [isOpen, setIsOpen] = useState()
  const hasInputMove = levelMoves.some((move) =>
    selectedMoveIDs.includes(move._id)
  )
  const color = hasInputMove ? 'var(--color-primary)' : 'var(--color-text)'
  const ulHeight = isOpen ? 'auto' : 0

  useEffect(() => {
    !selectedMoveIDs.length && setIsOpen(false)
  }, [selectedMoveIDs])

  function toogleInputLevel() {
    setIsOpen(!isOpen)
  }

  return (
    <InputLevelStyled isOpen={isOpen} color={color} ulHeight={ulHeight}>
      <h3 onClick={toogleInputLevel}>
        {isOpen ? <ArrowDownIcon /> : <ArrowRightIcon />}{' '}
        <span className="level-name">{levelName}</span>
      </h3>
      <ul>
        {levelMoves.map((move) => (
          <li key={move._id}>
            <label
              className={selectedMoveIDs.includes(move._id) ? 'isChecked' : ''}
            >
              <input
                name="move"
                value={move._id}
                onChange={updateAppState}
                type="checkbox"
                checked={selectedMoveIDs.includes(move.id)}
              />
              {move.name}
            </label>
          </li>
        ))}
      </ul>
    </InputLevelStyled>
  )
}

const InputLevelStyled = styled.div`
  background-color: var(--color-bg-accordion);
  border-radius: 5px;

  h3 {
    cursor: pointer;
    display: flex;
    justify-content: space-between;
    padding: 7px;
    font-size: 1rem;

    & .level-name {
      width: 100%;
      text-transform: uppercase;
      text-align: center;
      color: ${({ color }) => color};
    }

    svg {
      transform: scale(1.5);
    }
  }

  ul {
    display: grid;
    gap: 3px;
    padding: 0 6px;
    overflow: hidden;
    height: ${({ ulHeight }) => ulHeight};

    li {
      line-height: 1.3;

      input {
        display: none;
      }

      &:first-of-type {
        padding-top: 1px;
      }

      &:last-of-type {
        padding-bottom: 10px;
      }

      label {
        width: 100%;
        display: inline-block;
        font-size: 0.875rem;

        &:hover {
          cursor: pointer;
        }
      }
    }

    .isChecked {
      color: var(--color-primary);
    }
  }
`
