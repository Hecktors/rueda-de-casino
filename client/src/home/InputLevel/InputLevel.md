```jsx
<InputLevel
  levelName="level"
  levelMoves={[
    {
      _id: 'm1',
      name: 'la prima',
      steps: 2,
      filename: 'la_prima.mp3',
      videoId: 'SFxy0cWwEtg',
      videoStart: '20',
    },
    {
      _id: 'm3',
      name: 'sombrero con una mano',
      steps: 2,
      filename: 'sombrero_con_una_mano.mp3',
      videoId: 'fAMaQ_3HEUE',
      videoStart: '1',
    },
  ]}
  selectedMoveIDs={[]}
  updateUserInput={() => {}}
/><br/>
<InputLevel
  levelName="level"
  levelMoves={[
    {
      _id: 'm1',
      name: 'la prima',
      steps: 2,
      filename: 'la_prima.mp3',
      videoId: 'SFxy0cWwEtg',
      videoStart: '20',
    },
    {
      _id: 'm3',
      name: 'sombrero con una mano',
      steps: 2,
      filename: 'sombrero_con_una_mano.mp3',
      videoId: 'fAMaQ_3HEUE',
      videoStart: '1',
    },
  ]}
  selectedMoveIDs={['m1', 'm3']}
/>
```
