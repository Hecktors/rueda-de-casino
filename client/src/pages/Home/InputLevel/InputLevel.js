import { useState, useEffect } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components/macro'
import { device } from '../../../styles/device'
import { ArrowRightIcon, ArrowDownIcon } from '../../../components/Icons/Icons'

InputLevel.propTypes = {
  levelName: PropTypes.string.isRequired,
  levelMoves: PropTypes.array.isRequired,
  selectedMoveIds: PropTypes.array.isRequired,
  updateAppState: PropTypes.func.isRequired,
}

export default function InputLevel({
  levelName,
  levelMoves,
  selectedMoveIds,
  updateAppState,
}) {
  const [isOpen, setIsOpen] = useState()
  const isDesktop = window.innerWidth >= 1024
  const hasInputMove = levelMoves.some((move) =>
    selectedMoveIds.includes(move._id)
  )
  const color =
    hasInputMove && !isDesktop ? 'var(--color-text-active)' : 'transparent'
  const ulHeight = isOpen ? 'auto' : 0
  const headerSize = isDesktop ? '1.2rem' : '0.875rem'

  useEffect(() => {
    isDesktop ? setIsOpen(true) : !selectedMoveIds.length && setIsOpen(false)
  }, [selectedMoveIds]) // eslint-disable-line react-hooks/exhaustive-deps

  function toogleInputLevel() {
    !isDesktop && setIsOpen(!isOpen)
  }

  return (
    <InputLevelStyled
      isOpen={isOpen}
      color={color}
      ulHeight={ulHeight}
      headerSize={headerSize}
    >
      <h2 onClick={toogleInputLevel}>
        {!isDesktop && (isOpen ? <ArrowDownIcon /> : <ArrowRightIcon />)}{' '}
        <span className="level-name">{levelName}</span>
        <span className="checkmark">âœ”</span>
      </h2>
      <ul>
        {levelMoves.map((move) => (
          <li key={move._id}>
            <label
              className={selectedMoveIds.includes(move._id) ? 'isChecked' : ''}
            >
              <input
                name="move"
                value={move._id}
                onChange={updateAppState}
                type="checkbox"
                checked={selectedMoveIds.includes(move.id)}
              />
              {move.name}
            </label>
          </li>
        ))}
      </ul>
    </InputLevelStyled>
  )
}

const InputLevelStyled = styled.div`
  background-color: var(--color-bg-accordion);
  border-radius: 5px;

  h2 {
    cursor: pointer;
    display: flex;
    justify-content: space-between;
    padding: 7px;
    font-size: ${({ headerSize }) => headerSize};

    .level-name {
      width: 100%;
      text-transform: uppercase;
      text-align: center;
      color: var(--color-text);
    }

    .checkmark {
      color: ${({ color }) => color};
    }
  }

  ul {
    display: grid;
    padding: 0 5px;
    overflow: hidden;
    height: ${({ ulHeight }) => ulHeight};

    li {
      line-height: 1.3;

      input {
        display: none;
      }

      &:first-of-type {
        padding-top: 5px;
      }

      &:last-of-type {
        padding-bottom: 15px;
      }

      label {
        height: 100%;
        padding: 0;
        width: 100%;
        display: inline-block;
        font-size: 0.875rem;

        @media ${device.laptop} {
          &:hover {
            cursor: pointer;
            opacity: 0.5;
          }
        }
      }
    }

    .isChecked {
      color: var(--color-text-active);
    }
  }
`
