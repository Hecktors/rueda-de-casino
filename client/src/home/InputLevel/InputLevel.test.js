import ReactDOM from 'react-dom'
import InputLevel from './InputLevel'
const moves = [
  {
    _id: 1,
    name: 'la prima',
    steps: 2,
    filename: 'la_prima.mp3',
  },
  {
    _id: 2,
    name: 'sombrero',
    steps: 2,
    filename: 'sombrero.mp3',
  },
]

describe('InputLevel', () => {
  it('renders Accordion', () => {
    const div = document.createElement('div')
    ReactDOM.render(
      <InputLevel
        name="test accordion"
        levelMoves={moves}
        levelName="level"
        selectedMoveIDs={[]}
        updateAppState={() => {}}
      />,
      div
    )
    expect(div.firstChild).toMatchSnapshot()
  })
})
