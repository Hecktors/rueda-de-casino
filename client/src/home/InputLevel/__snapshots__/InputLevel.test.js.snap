// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InputLevel renders Accordion 1`] = `
<div
  class="InputLevel__InputLevelStyled-g2yagl-0 fOwyHI"
  color="var(--color-text)"
>
  <h3>
    <svg>
      arrow_right.svg
    </svg>
     
    <span
      class="level-name"
    >
      level
    </span>
  </h3>
  <ul>
    <li>
      <label
        class=""
      >
        <input
          name="move"
          type="checkbox"
          value="1"
        />
        la prima
      </label>
    </li>
    <li>
      <label
        class=""
      >
        <input
          name="move"
          type="checkbox"
          value="2"
        />
        sombrero
      </label>
    </li>
  </ul>
</div>
`;
